// @dcloudio/uts/dist/cli.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cac_1 = require("cac");
const index_1 = require("./index");
const cli = (0, cac_1.cac)('uts');
cli
    .command('<input> [output]')
    .option('-t, --target <target>', '[string] kotlin | swift', {
    default: index_1.UtsTarget.KOTLIN,
})
    .option('-s, --sourceMap [sourceMap]', `[boolean|string] output sourceMap`, {
    default: false,
})
    .option('-i, --inlineSourcesContent [inlineSourcesContent]', `[boolean] inline sources content`, {
    default: false,
})
    .option('-w, --watch', `[boolean] rebuilds when uts have changed on disk`, {
    default: false,
})
    .option('-e, --extname [extname]', `[string] extname`, {
    default: '.uts',
})
    .option('--silent', `[boolean] disable log`, {
    default: false,
})
    .action((input, output, opts) => {
    const toOptions = {
        silent: opts.silent,
        input: {
            dir: input,
            extname: opts.extname,
        },
        output: {
            dir: output,
            sourceMap: opts.sourceMap,
            inlineSourcesContent: opts.inlineSourcesContent,
            extname: index_1.UtsTargetExtNames[opts.target],
        },
    };
    return opts.watch
        ? (0, index_1.runDev)(opts.target, toOptions)
        : (0, index_1.runBuild)(opts.target, toOptions);
});
cli.help();
cli.version(require('../package.json').version);
cli.parse();
